### PR TITLE
Support for percentage based vaccination details

### PR DESCRIPTION
**Pull request by handyjellyfish**
Adds support for percentages of population for first and second dose. 
Falls back to previous updates vaccinations if latest update does not contain them.
Falls back to latest update if latest record is null

### CODE REVIEW

**Comment by handyjellyfish** [comment]
Population data matches what Aus Gov uses for their calculations. Possibly could find a dynamic data source, but unlikely to change much for the lifetime of this bot. Numbers sourced from this document: https://www.health.gov.au/sites/default/files/documents/2021/09/covid-19-vaccine-rollout-update-8-september-2021.pdf

**Comment by handyjellyfish** [comment]
I want to add a Vax Stats details - but will require a bit of data manipulation to get percentages so that is tonights job :D

File: post_covid_stats.py [hunk 1]
```diff
@@ -6,23 +6,29 @@
...
 
 def post_covid_stats() -> SlackResponse:
+    population_data_file: str = "resources/population.json"
     source_url: str = "https://covidlive.com.au/covid-live.json"
     covid_data: Dict = fetch_and_parse_data(source_url)
-
+    
+    with open(population_data_file) as populationDataFile:
+        population_data: Dict = json.load(populationDataFile);
+        populationDataFile.close();
```
Comment (B0yc3y) [inline post_covid_stats.py]: Shouldn't need to explicitly close when using a with block IIRC

**Review by B0yc3y — CHANGES_REQUESTED** [review]
Looks good. Could you please update the test that I now remembered to check in

**Comment by B0yc3y** [comment]
> I want to add a Vax Stats details - but will require a bit of data manipulation to get percentages so that is tonight's job :D

I also started this work in the background. Here is as far as I got. 

```
        # New vaccinations in reporting period
        new_vax: int = None
        if code_data['VACC_DOSE_CNT'] is not None:
            new_vax = int(code_data['VACC_DOSE_CNT']) - int(code_data['PREV_VACC_DOSE_CNT'])

        # New first dose in reporting period
        first_dose_vax_change: int = None
        if code_data['VACC_FIRST_DOSE_CNT'] is not None:
            first_dose_vax_change = int(code_data['VACC_FIRST_DOSE_CNT']) - int(code_data['PREV_VACC_FIRST_DOSE_CNT'])

        # New filly vaxed in reporting period
        new_fully_vax: int = None
        if code_data['VACC_PEOPLE_CNT'] is not None:
            new_fully_vax = int(code_data['VACC_PEOPLE_CNT']) - int(code_data['PREV_VACC_PEOPLE_CNT'])

---------------------------------------

            if code_data["VACC_FIRST_DOSE_CNT"] is not None and code_data["VACC_FIRST_DOSE_CNT"] != 0:
                pop_percent = self.population_percent(int(code_data['VACC_FIRST_DOSE_CNT']), population)
                message += f" | {format(pop_percent, '.2f')} 1st dose"

                if first_dose_vax_change is not None:
                    change_percent = self.population_percent(first_dose_vax_change, population)
                    message += f" ({'+' if int(change_percent) >= 0 else ''}{format(change_percent, '.2f')}%)"

            if code_data["VACC_PEOPLE_CNT"] is not None and code_data["VACC_PEOPLE_CNT"] != 0:
                pop_percent = self.population_percent(int(code_data['VACC_PEOPLE_CNT']), population)
                message += f" | {format(pop_percent, '.2f')} Fully Vaxed \n"

                if new_fully_vax is not None:
                    change_percent = self.population_percent(new_fully_vax, population)
                    message += f" ({'+' if int(change_percent) >= 0 else ''}{format(change_percent, '.2f')}%)\n"


-----------------------

    def population_percent(self, count: int, population: int) -> float:
        return count / population * 100
```

**Comment by B0yc3y** [comment]
> Population data matches what Aus Gov uses for their calculations. Possibly could find a dynamic data source, but unlikely to change much for the lifetime of this bot. Numbers sourced from this document: https://www.health.gov.au/sites/default/files/documents/2021/09/covid-19-vaccine-rollout-update-8-september-2021.pdf

Agreed, static data is fine in this case. I was taking the same approach.

File: post_covid_stats.py (hunk 1 above)
Comment (handyjellyfish) [reply post_covid_stats.py]: My python foo is lacking, as you can see.

**Comment by handyjellyfish** [comment]
> > I want to add a Vax Stats details - but will require a bit of data manipulation to get percentages so that is tonight's job :D
> 
> I also started this work in the background. Here is as far as I got.
> 
> ```
>         # New vaccinations in reporting period
>         new_vax: int = None
>         if code_data['VACC_DOSE_CNT'] is not None:
>             new_vax = int(code_data['VACC_DOSE_CNT']) - int(code_data['PREV_VACC_DOSE_CNT'])
> 
>         # New first dose in reporting period
>         first_dose_vax_change: int = None
>         if code_data['VACC_FIRST_DOSE_CNT'] is not None:
>             first_dose_vax_change = int(code_data['VACC_FIRST_DOSE_CNT']) - int(code_data['PREV_VACC_FIRST_DOSE_CNT'])
> 
>         # New filly vaxed in reporting period
>         new_fully_vax: int = None
>         if code_data['VACC_PEOPLE_CNT'] is not None:
>             new_fully_vax = int(code_data['VACC_PEOPLE_CNT']) - int(code_data['PREV_VACC_PEOPLE_CNT'])
> 
> ---------------------------------------
> 
>             if code_data["VACC_FIRST_DOSE_CNT"] is not None and code_data["VACC_FIRST_DOSE_CNT"] != 0:
>                 pop_percent = self.population_percent(int(code_data['VACC_FIRST_DOSE_CNT']), population)
>                 message += f" | {format(pop_percent, '.2f')} 1st dose"
> 
>                 if first_dose_vax_change is not None:
>                     change_percent = self.population_percent(first_dose_vax_change, population)
>                     message += f" ({'+' if int(change_percent) >= 0 else ''}{format(change_percent, '.2f')}%)"
> 
>             if code_data["VACC_PEOPLE_CNT"] is not None and code_data["VACC_PEOPLE_CNT"] != 0:
>                 pop_percent = self.population_percent(int(code_data['VACC_PEOPLE_CNT']), population)
>                 message += f" | {format(pop_percent, '.2f')} Fully Vaxed \n"
> 
>                 if new_fully_vax is not None:
>                     change_percent = self.population_percent(new_fully_vax, population)
>                     message += f" ({'+' if int(change_percent) >= 0 else ''}{format(change_percent, '.2f')}%)\n"
> 
> 
> -----------------------
> 
>     def population_percent(self, count: int, population: int) -> float:
>         return count / population * 100
> ```

Yeah - I think that covidlive uses a 7 day rolling average to calculate this - so was thinking of trying something similar. Would likely get the last seven days of data. And calc on that. I think that we can safely assume the data is in date order, so wouldn't be too onerous to pick it out.

**Review by B0yc3y — APPROVED** [review]
Nice one. LGTM 👍

**Pull request merged**